gapic/src/main/com/google/gapid/perfetto/models/CounterTrack.java

```diff
@@ -158,44 +167,46 @@ private String rangeSql(TimeSpan ts) {
   public static class Data extends Track.Data {
     public final long[] ts;
     public final double[] values;
+    public final long[] ids;
 
-    public Data(DataRequest request, long[] ts, double[] values) {
+    public Data(DataRequest request, long[] ts, double[] values, long[] ids) {
       super(request);
       this.ts = ts;
       this.values = values;
+      this.ids = ids;
     }
 
     public static Data empty(DataRequest req) {
-      return new Data(req, new long[0], new double[0]);
+      return new Data(req, new long[0], new double[0], new long[0]);
     }
   }
 
-  public static class Values implements Selection<Values.Key>, Selection.Builder<Values> {
+  public static class Values implements Selection<Long>, Selection.Builder<Values> {
     public final long[] ts;
     public final String[] names;
     public final double[][] values;
-    private final Set<Values.Key> valueKeys = Sets.newHashSet();
+    public final long[][] ids;
+    private final Set<Long> valueKeys = Sets.newHashSet();
 
     public Values(String name, Data data) {
       this.ts = data.ts;
       this.names = new String[] { name };
       this.values = new double[][] { data.values };
+      this.ids = new long[][] { data.ids };
       initKeys();
     }
 
-    private Values(long[] ts, String[] names, double[][] values) {
+    private Values(long[] ts, String[] names, double[][] values, long[][] ids) {
       this.ts = ts;
       this.names = names;
       this.values = values;
+      this.ids = ids;
       initKeys();
     }
 
     private void initKeys() {
-      for (String name : names) {
-        // Skip the last dummy entry for the range end.
-        Arrays.stream(ts, 0, ts.length - 1)
-            .boxed()
-            .forEach(t -> valueKeys.add(new Values.Key(name, t)));
+      for (long[] keys : ids) {
+        valueKeys.addAll(Arrays.stream(keys).boxed().collect(Collectors.toSet()));
```

Comment:
Nit: Didn't notice this before. The old code seems to ignore the last dummy entry. I. don't think the new one does that. However, I'm not too sure of whether it will affect the end result or not.

Reply:
It's probably better this way in the new code.  The old code use name + ts as the key. The last value is artificially created and doesn't actually exist in the database.  In the new code, the key is the id.  The last value is just a duplicate of the previous value.  Since it's being inserted into a Set, the duplicate just get ignored.

Reply:
Yes, it's taking advantage of set+long to keep code shorter.